packages/overture-schema-system/src/overture/schema/system/json_schema.py

```diff
@@ -0,0 +1,204 @@
+    {
+      "properties": {
+        "foo": {
+          "anyOf": [
+            {
+              "type": "string"
+            },
+            {
+              "type": "null"
+            }
+          ],
+          "default": null,
+          "title": "Foo"
+        }
+      },
+      "title": "FooModel",
+      "type": "object"
+    }
+
+    This default Pydantic behavior is reasonable, and aligns well with the Python internal
+    representation. However, it doesn't take advantage of a quiet advantage of JSON and JSON Schema,
+    which is that there's another concept of optionality that does not require the value `null` at
+    all. In this paradigm, if the field is there, it has a set value; and if the field is missing,
+    it does not. Pydantic does enable this alternative paradigm via the experimental `MISSING`
+    sentinel, but this sentinel is not yet a full-fledged feature and has some rough edges.
+
+    The purpose of this class is to generate a JSON Schema that aligns with the alternative paradigm
+    where nullable optional fields are omitted from the JSON rather than assigned the literal value
+    `null`.
+
+    Transformations
+    ---------------
+    This class makes the following changes to the default Pydantic JSON Schema for nullable optional
+    fields only:
+
+    1. The JSON Schema type `null` is removed as an allowed type from the field's schema, which will
+       usually result in eliminating the `"anyOf"` composition keyword from the field's schema.
+    2. If the field's default value is `None` (JSON `null`), the default value is removed.
+    """
+
+    @override
+    def default_schema(self, schema: core_schema.WithDefaultSchema) -> JsonSchemaValue:
+        match GenerateOmitNullableOptionalJsonSchema._redact_nullable_schema(schema):
+            case redacted if redacted and schema["default"] is None:
+                return self.generate_inner(redacted["schema"])
+            case redacted if redacted:
+                return self.generate_inner(redacted)
+            case _:
+                return super().default_schema(schema)
+
+    @staticmethod
+    def _redact_nullable_schema(
+        schema: core_schema.CoreSchema,
+    ) -> core_schema.CoreSchema | None:
+        match schema.get("schema", None):
+            case sub_schema if sub_schema and sub_schema["type"] == "nullable":
+                return {**schema, "schema": sub_schema["schema"]}
+            case sub_schema if sub_schema:
+                redacted = (
+                    GenerateOmitNullableOptionalJsonSchema._redact_nullable_schema(
+                        sub_schema
+                    )
+                )
+                return None if not redacted else {**schema, "schema": redacted}
+            case _:
+                return None
+
+
+def json_schema(thing: object) -> JsonSchemaValue:
+    """
+    Generate JSON Schema for a Pydantic model or union of models.
```

Comment:
YAGNI, but if we're truly being thorough, shouldn't this include support for dataclasses?